xbmc/network/WebServer.cpp

```diff
@@ -599,6 +695,68 @@ void CWebServer::FinalizePostDataProcessing(ConnectionHandler *connectionHandler
   MHD_destroy_post_processor(connectionHandler->postprocessor);
 }
 
+int CWebServer::HandleOptionsRequest(HTTPRequest request)
+{
+  // make sure the requested method is supported (HEAD, GET, POST)
+  const std::string headerMethod = HTTPRequestHandlerUtils::GetRequestHeaderValue(request.connection, MHD_HEADER_KIND, HTTP_HEADER_ACCESS_CONTROL_REQUEST_METHOD);
+  auto meth = GetHTTPMethod(headerMethod.c_str());
+  if (SUPPORTED_HTTP_METHODS.find(meth) == SUPPORTED_HTTP_METHODS.end())
+  {
+    CLog::Log(LOGWARNING, "CWebServer[%hu]: bad OPTIONS request with unsupported \"%s\" header: %s",
+      m_port, HTTP_HEADER_ACCESS_CONTROL_REQUEST_METHOD.c_str(), headerMethod.c_str());
+    return SendErrorResponse(request, MHD_HTTP_BAD_REQUEST, OPTIONS);
+  }
+
+  // create a fake potential request and check if it can be handled
+  HTTPRequest potentialRequest = request;
+  potentialRequest.method = meth;
+  auto requestHandler = FindRequestHandler(potentialRequest);
+  if (requestHandler == nullptr)
+  {
+    CLog::Log(LOGERROR, "CWebServer[%hu]: couldn't find a potential request handler for %s", m_port, request.pathUrl.c_str());
+    return SendErrorResponse(request, MHD_HTTP_NOT_FOUND, request.method);
+  }
+
+  struct MHD_Response *response = create_response(0, nullptr, MHD_NO, MHD_NO);
+  if (response == nullptr)
+  {
+    CLog::Log(LOGERROR, "CWebServer[%hu]: failed to create a HTTP 200 OPTIONS response", m_port);
+    return MHD_NO;
```

Comment:
CORS headers needs to be added to all non-ok responses as well, otherwise the client will be unable to read the error response.